include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -947,15 +954,12 @@ FEPointEvaluation<n_components, dim, spacedim, Number>::evaluate(
                     Number>::set_gradient(val_and_grad.second,
                                           j,
                                           unit_gradients[i + j]);
-                  gradients[i + j] =
-                    static_cast<typename internal::FEPointEvaluation::
-                                  EvaluatorTypeTraits<dim,
-                                                      n_components,
-                                                      Number>::gradient_type>(
+                  gradients[i + j] = static_cast<gradient_type>(
+                    static_cast<gradient_type_double>(
                       apply_transformation(mapping_info->get_mapping_data()
                                              .inverse_jacobians[i + j]
                                              .transpose(),
-                                           unit_gradients[i + j]));
+                                           unit_gradients[i + j])));
```

Comment:
Trying to understand this issue, I think that the fix here works around the type that gets set here:
https://github.com/dealii/dealii/blob/d1475a1c001e1bffe5bfbcc6fb804df8868e371b/include/deal.II/base/derivative_form.h#L481-L496
Is this a correct observation @bergbauer @peterrum ? In that case, are we sure that the type of `apply_transformation` is the most appropriate one, or should we rather change it to pick `Number2`, or create a variant that always picks the desired type `Number2`?

Reply:
The problematic one is https://github.com/dealii/dealii/blob/d1475a1c001e1bffe5bfbcc6fb804df8868e371b/include/deal.II/base/derivative_form.h#L451-L465
which is why we need a cast from `DerivativeForm` to `Tensor` with lines https://github.com/dealii/dealii/blob/d1475a1c001e1bffe5bfbcc6fb804df8868e371b/include/deal.II/base/derivative_form.h#L287-L300

There are three options in my optinion:

1. Let all `apply_transformation` functions return `Tensor` objects. Then cast from `double` to `float` or teach it to the assignment operator.
2. Leave `apply_transformation` as is. Fix the `Number` return type of `assign_transformation` to `Number2`. Then cast from `DerivativeForm` to `Tensor` (your suggestion).
3. First cast from `DerivativeForm` to `Tensor<double>` and then cast from `Tensor<double>` to `Tensor<float>` or teach it to the assignment operator.

Reply:
What do you think of an overload
```cpp
template <int dim, typename Number1, typename Number2>
inline Tensor<2, dim, typename ProductType<Number1, Number2>::type>
apply_transformation(const DerivativeForm<1, dim, dim, Number1> &grad_F,
                     const Tensor<2, dim, Number2> &             D_X)
{
  return apply_transformation(grad_F, D_X);
}
```
?
This is basically option 1.

Reply:
I went with option 1 in the last commit https://github.com/dealii/dealii/pull/14159/commits/0c6c42f979c49f4ec1622e8c57d452dc8e9e1578 . The failing assignment for nested Tensors with different `Number` type was nasty to find, see the new test.